File: sdks/java/src/main/java/io/linktor/resources/VREResource.java
package io.linktor.resources;

import io.linktor.types.VRE;
import io.linktor.utils.HttpClient;
import io.linktor.utils.LinktorException;

import java.util.HashMap;
import java.util.Map;

/**
 * VRE (Visual Response Engine) Resource
 *
 * Render visual templates as images for messaging channels.
 */
public class VREResource {
    private final HttpClient http;

    public VREResource(HttpClient http) {
        this.http = http;
    }

    /**
     * Render a VRE template to an image.
     * Returns base64-encoded image data that can be sent to messaging channels.
     */
    public VRE.RenderResponse render(VRE.RenderRequest request) throws LinktorException {
        return http.post("/vre/render", request, VRE.RenderResponse.class);
    }

    /**
     * Render a VRE template and send it directly to a conversation.
     * Combines rendering and sending in one operation.
     */
    public VRE.RenderAndSendResponse renderAndSend(VRE.RenderAndSendRequest request) throws LinktorException {
        return http.post("/vre/render-and-send", request, VRE.RenderAndSendResponse.class);
    }

    /**
     * List available VRE templates with their schemas and example data.
     */
    public VRE.ListTemplatesResponse listTemplates() throws LinktorException {
        return listTemplates(null);
    }

    /**
     * List available VRE templates with their schemas and example data.
     *
     * @param tenantId Optional tenant ID to include custom templates
     */
    public VRE.ListTemplatesResponse listTemplates(String tenantId) throws LinktorException {
        Map<String, String> queryParams = new HashMap<>();
        if (tenantId != null && !tenantId.isEmpty()) {
            queryParams.put("tenant_id", tenantId);
        }
        return http.get("/vre/templates", queryParams, VRE.ListTemplatesResponse.class);
    }

    /**
     * Preview a VRE template with sample data.
     *
     * @param templateId Template to preview
     * @param data Optional custom data (uses default sample data if not provided)
     */
    public VRE.PreviewResponse preview(String templateId, Map<String, Object> data) throws LinktorException {
        VRE.PreviewRequest request = new VRE.PreviewRequest(data);
        return http.post("/vre/templates/" + templateId + "/preview", request, VRE.PreviewResponse.class);
    }

    /**
     * Preview a VRE template with default sample data.
     */
    public VRE.PreviewResponse preview(String templateId) throws LinktorException {
        return preview(templateId, null);
    }

    // ============================================
    // Convenience methods for common templates
    // ============================================

    /**
     * Render a menu with numbered options.
     *
     * @param tenantId Tenant ID
     * @param titulo Menu title
     * @param opcoes List of options (label, icone, descricao)
     * @param channel Target channel
     */
    public VRE.RenderResponse renderMenu(String tenantId, String titulo, java.util.List<Map<String, Object>> opcoes, VRE.ChannelType channel) throws LinktorException {
        Map<String, Object> data = new HashMap<>();
        data.put("titulo", titulo);
        data.put("opcoes", opcoes);

        return render(VRE.RenderRequest.builder()
                .tenantId(tenantId)
                .templateId(VRE.TemplateType.MENU_OPCOES)
                .data(data)
                .channel(channel)
                .build());
    }

    /**
     * Render a product card.
     *
     * @param tenantId Tenant ID
     * @param nome Product name
     * @param preco Product price
     * @param unidade Unit (kg, un, cx, etc.)
     * @param channel Target channel
     */
    public VRE.RenderResponse renderProductCard(String tenantId, String nome, double preco, String unidade, VRE.ChannelType channel) throws LinktorException {
        Map<String, Object> data = new HashMap<>();
        data.put("nome", nome);
        data.put("preco", preco);
        data.put("unidade", unidade);

        return render(VRE.RenderRequest.builder()
                .tenantId(tenantId)
                .templateId(VRE.TemplateType.CARD_PRODUTO)
                .data(data)
                .channel(channel)
                .build());
    }

    /**
     * Render an order status timeline.
     *
     * @param tenantId Tenant ID
     * @param numeroPedido Order number
     * @param statusAtual Current status
     * @param channel Target channel
     */
    public VRE.RenderResponse renderOrderStatus(String tenantId, String numeroPedido, VRE.OrderStatus statusAtual, VRE.ChannelType channel) throws LinktorException {
        Map<String, Object> data = new HashMap<>();
        data.put("numero_pedido", numeroPedido);
        data.put("status_atual", statusAtual.name().toLowerCase());

        return render(VRE.RenderRequest.builder()
                .tenantId(tenantId)
                .templateId(VRE.TemplateType.STATUS_PEDIDO)
                .data(data)
                .channel(channel)
                .build());
    }

    /**
     * Render a product list for comparison.
     *
     * @param tenantId Tenant ID
     * @param titulo List title
     * @param produtos List of products
     * @param channel Target channel
     */
    public VRE.RenderResponse renderProductList(String tenantId, String titulo, java.util.List<Map<String, Object>> produtos, VRE.ChannelType channel) throws LinktorException {
        Map<String, Object> data = new HashMap<>();
        data.put("titulo", titulo);
        data.put("produtos", produtos);

        return render(VRE.RenderRequest.builder()
                .tenantId(tenantId)
                .templateId(VRE.TemplateType.LISTA_PRODUTOS)
                .data(data)
                .channel(channel)
                .build());
    }

    /**
     * Render a confirmation summary.
     *
     * @param tenantId Tenant ID
     * @param valorTotal Total value
     * @param itens List of items
     * @param channel Target channel
     */
    public VRE.RenderResponse renderConfirmation(String tenantId, double valorTotal, java.util.List<Map<String, Object>> itens, VRE.ChannelType channel) throws LinktorException {
        Map<String, Object> data = new HashMap<>();
        data.put("valor_total", valorTotal);
        data.put("itens", itens);

        return render(VRE.RenderRequest.builder()
                .tenantId(tenantId)
                .templateId(VRE.TemplateType.CONFIRMACAO)
                .data(data)
                .channel(channel)
                .build());
    }

    /**
     * Render a PIX payment QR code.
     *
     * @param tenantId Tenant ID
     * @param valor Payment amount
     * @param pixPayload PIX EMV/BRCode payload
     * @param channel Target channel
     */
    public VRE.RenderResponse renderPixPayment(String tenantId, double valor, String pixPayload, VRE.ChannelType channel) throws LinktorException {
        Map<String, Object> data = new HashMap<>();
        data.put("valor", valor);
        data.put("pix_payload", pixPayload);

        return render(VRE.RenderRequest.builder()
                .tenantId(tenantId)
                .templateId(VRE.TemplateType.COBRANCA_PIX)
                .data(data)
                .channel(channel)
                .build());
    }
}

File: sdks/php/src/Resources/VREResource.php
<?php

declare(strict_types=1);

namespace Linktor\Resources;

use Linktor\LinktorClient;
use Linktor\Types\VRERenderResponse;
use Linktor\Types\VRERenderAndSendResponse;
use Linktor\Types\VREListTemplatesResponse;
use Linktor\Types\VREPreviewResponse;

/**
 * VRE (Visual Response Engine) Resource
 *
 * Render visual templates as images for messaging channels.
 */
class VREResource
{
    private LinktorClient $client;

    public function __construct(LinktorClient $client)
    {
        $this->client = $client;
    }

    /**
     * Render a VRE template to an image.
     * Returns base64-encoded image data that can be sent to messaging channels.
     *
     * @param string $tenantId Tenant ID
     * @param string $templateId Template to render (menu_opcoes, card_produto, etc.)
     * @param array $data Template data
     * @param string|null $channel Target channel (whatsapp, telegram, web, email)
     * @param string|null $format Output format (png, webp, jpeg)
     * @param int|null $width Override width
     * @param int|null $quality Quality 0-100 (for webp/jpeg)
     * @param float|null $scale Scale factor (1.0-2.0)
     */
    public function render(
        string $tenantId,
        string $templateId,
        array $data,
        ?string $channel = null,
        ?string $format = null,
        ?int $width = null,
        ?int $quality = null,
        ?float $scale = null
    ): VRERenderResponse {
        $payload = [
            'tenant_id' => $tenantId,
            'template_id' => $templateId,
            'data' => $data,
        ];

        if ($channel !== null) $payload['channel'] = $channel;
        if ($format !== null) $payload['format'] = $format;
        if ($width !== null) $payload['width'] = $width;
        if ($quality !== null) $payload['quality'] = $quality;
        if ($scale !== null) $payload['scale'] = $scale;

        $response = $this->client->post('/vre/render', $payload);
        return VRERenderResponse::fromArray($response);
    }

    /**
     * Render a VRE template and send it directly to a conversation.
     *
     * @param string $conversationId Conversation ID to send to
     * @param string $templateId Template to render
     * @param array $data Template data
     * @param string|null $caption Optional custom caption
     * @param string|null $followUpText Optional text to send after image
     */
    public function renderAndSend(
        string $conversationId,
        string $templateId,
        array $data,
        ?string $caption = null,
        ?string $followUpText = null
    ): VRERenderAndSendResponse {
        $payload = [
            'conversation_id' => $conversationId,
            'template_id' => $templateId,
            'data' => $data,
        ];

        if ($caption !== null) $payload['caption'] = $caption;
        if ($followUpText !== null) $payload['follow_up_text'] = $followUpText;

        $response = $this->client->post('/vre/render-and-send', $payload);
        return VRERenderAndSendResponse::fromArray($response);
    }

    /**
     * List available VRE templates.
     *
     * @param string|null $tenantId Optional tenant ID to include custom templates
     */
    public function listTemplates(?string $tenantId = null): VREListTemplatesResponse
    {
        $path = '/vre/templates';
        if ($tenantId !== null) {
            $path .= '?tenant_id=' . urlencode($tenantId);
        }

        $response = $this->client->get($path);
        return VREListTemplatesResponse::fromArray($response);
    }

    /**
     * Preview a VRE template with sample data.
     *
     * @param string $templateId Template to preview
     * @param array|null $data Optional custom data (uses defaults if not provided)
     */
    public function preview(string $templateId, ?array $data = null): VREPreviewResponse
    {
        $payload = $data !== null ? ['data' => $data] : [];
        $response = $this->client->post("/vre/templates/{$templateId}/preview", $payload);
        return VREPreviewResponse::fromArray($response);
    }

    // ============================================
    // Convenience methods for common templates
    // ============================================

    /**
     * Render a menu with numbered options.
     *
     * @param string $tenantId Tenant ID
     * @param string $titulo Menu title
     * @param array $opcoes List of options (with label, icone, descricao)
     * @param string $channel Target channel
     */
    public function renderMenu(
        string $tenantId,
        string $titulo,
        array $opcoes,
        string $channel = 'whatsapp'
    ): VRERenderResponse {
        return $this->render($tenantId, 'menu_opcoes', [
            'titulo' => $titulo,
            'opcoes' => $opcoes,
        ], $channel);
    }

    /**
     * Render a product card.
     *
     * @param string $tenantId Tenant ID
     * @param string $nome Product name
     * @param float $preco Product price
     * @param string $unidade Unit (kg, un, cx, etc.)
     * @param string $channel Target channel
     */
    public function renderProductCard(
        string $tenantId,
        string $nome,
        float $preco,
        string $unidade,
        string $channel = 'whatsapp'
    ): VRERenderResponse {
        return $this->render($tenantId, 'card_produto', [
            'nome' => $nome,
            'preco' => $preco,
            'unidade' => $unidade,
        ], $channel);
    }

    /**
     * Render an order status timeline.
     *
     * @param string $tenantId Tenant ID
     * @param string $numeroPedido Order number
     * @param string $statusAtual Current status (recebido, separacao, faturado, transporte, entregue)
     * @param string $channel Target channel
     */
    public function renderOrderStatus(
        string $tenantId,
        string $numeroPedido,
        string $statusAtual,
        string $channel = 'whatsapp'
    ): VRERenderResponse {
        return $this->render($tenantId, 'status_pedido', [
            'numero_pedido' => $numeroPedido,
            'status_atual' => $statusAtual,
        ], $channel);
    }

    /**
     * Render a product list for comparison.
     *
     * @param string $tenantId Tenant ID
     * @param string $titulo List title
     * @param array $produtos List of products
     * @param string $channel Target channel
     */
    public function renderProductList(
        string $tenantId,
        string $titulo,
        array $produtos,
        string $channel = 'whatsapp'
    ): VRERenderResponse {
        return $this->render($tenantId, 'lista_produtos', [
            'titulo' => $titulo,
            'produtos' => $produtos,
        ], $channel);
    }

    /**
     * Render a confirmation summary.
     *
     * @param string $tenantId Tenant ID
     * @param float $valorTotal Total value
     * @param array $itens List of items
     * @param string $channel Target channel
     */
    public function renderConfirmation(
        string $tenantId,
        float $valorTotal,
        array $itens,
        string $channel = 'whatsapp'
    ): VRERenderResponse {
        return $this->render($tenantId, 'confirmacao', [
            'valor_total' => $valorTotal,
            'itens' => $itens,
        ], $channel);
    }

    /**
     * Render a PIX payment QR code.
     *
     * @param string $tenantId Tenant ID
     * @param float $valor Payment amount
     * @param string $pixPayload PIX EMV/BRCode payload
     * @param string $channel Target channel
     */
    public function renderPixPayment(
        string $tenantId,
        float $valor,
        string $pixPayload,
        string $channel = 'whatsapp'
    ): VRERenderResponse {
        return $this->render($tenantId, 'cobranca_pix', [
            'valor' => $valor,
            'pix_payload' => $pixPayload,
        ], $channel);
    }
}
